main/autogen/autogen.cc

```diff
@@ -663,15 +664,116 @@ string ParsedFile::toMsgpack(core::Context ctx, int version) {
     return write.pack(ctx, *this);
 }
 
-void ParsedFile::classlist(core::Context ctx, vector<string> &out) {
-    auto nameToString = [&](const auto &nm) -> string { return nm.data(ctx)->show(ctx); };
+vector<string> ParsedFile::listAllClasses(const core::Context ctx) {
+    vector<string> out;
+
     for (auto &def : defs) {
         if (def.type != Definition::Class) {
             continue;
         }
         auto names = showFullName(ctx, def.id);
-        out.emplace_back(fmt::format("{}", fmt::map_join(names, "::", nameToString)));
+        out.emplace_back(fmt::format("{}", fmt::map_join(names, "::", [&ctx](const core::NameRef &nm) -> string {
+                                         return nm.data(ctx)->show(ctx);
+                                     })));
+    }
+
+    return out;
+}
+
+optional<AutogenSubclassMap> ParsedFile::listAllSubclasses(const core::Context ctx,
+                                                           const vector<string> &absolutePathsToIgnore,
+                                                           const vector<string> &relativePathsToIgnore) {
+    // We prepend "/" to `path` to mimic how `isFileIgnored` gets called elsewhere
+    if (sorbet::FileOps::isFileIgnored("", fmt::format("/{}", path), absolutePathsToIgnore, relativePathsToIgnore)) {
+        return nullopt;
+    }
+
+    AutogenSubclassMap out;
+
+    for (const Reference &ref : refs) {
+        DefinitionRef defn = ref.parent_of;
+
+        if (!defn.exists()) {
+            continue;
+        }
+
+        // Get fully-qualified parent name as string
+        string parentName =
+            fmt::format("{}", fmt::map_join(ref.resolved, "::", [&ctx](const core::NameRef &nm) -> string {
+                            return nm.data(ctx)->show(ctx);
+                        }));
+
+        // Add child class to the set identified by its parent
+        string childName =
+            fmt::format("{}", fmt::map_join(showFullName(ctx, defn), "::", [&ctx](const core::NameRef &nm) -> string {
+                            return nm.data(ctx)->show(ctx);
+                        }));
+
+        out[parentName].insert(make_pair(childName, defn.data(*this).type));
+    }
+
+    return out;
+}
+
+// Generate all descendants of a parent class
+// Recursively walks `childMap`, which stores the IMMEDIATE children of subclassed class.
+void descendantsOf(const AutogenSubclassMap &childMap, const string &parentName, AutogenSubclassSet &out) {
```

Comment:
please return a value rather than using an out parameter for the externally exposed method

Reply:
[fixed](https://github.com/sorbet/sorbet/pull/1040/files#diff-fe69bef0ea7ee4e25c450060a16165f5R70)